src/Microsoft.AspNetCore.Server.Testing/Deployers/NginxDeployer.cs

```diff
@@ -0,0 +1,138 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Net.Http;
+using Microsoft.AspNetCore.Server.Testing.Common;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.Server.Testing
+{
+    /// <summary>
+    /// Deployer for Kestrel on Nginx.
+    /// </summary>
+    public class NginxDeployer : SelfHostDeployer
+    {
+        private string configFile;
+        private readonly int waitTime = (int)TimeSpan.FromSeconds(2).TotalMilliseconds;
+
+        public NginxDeployer(DeploymentParameters deploymentParameters, ILogger logger)
+            : base(deploymentParameters, logger)
+        {
+        }
+
+        public override DeploymentResult Deploy()
+        {
+            configFile = Path.Combine(DeploymentParameters.ApplicationPath, Path.GetTempFileName());
+            var uri = new Uri(DeploymentParameters.ApplicationBaseUriHint);
+
+            var redirectUri = $"http://localhost:{TestUriHelper.FindFreePort()}";
+
+            var exitToken = StartSelfHost(new Uri(redirectUri));
+
+            SetupNginx(redirectUri, uri);
+
+            // Wait for App to be loaded since Nginx returns 502 instead of 503 when App isn't loaded
+            // Target actual address to avoid going through Nginx proxy
+            using (var httpClient = new HttpClient())
+            {
+                var response = RetryHelper.RetryRequest(() =>
+                {
+                    return httpClient.GetAsync(redirectUri);
+                }, Logger).Result;
+
+                if (!response.IsSuccessStatusCode)
+                {
+                    throw new InvalidOperationException("Deploy failed");
+                }
+            }
+            
+            return new DeploymentResult
+            {
+                WebRootLocation = DeploymentParameters.ApplicationPath,
+                DeploymentParameters = DeploymentParameters,
+                ApplicationBaseUri = uri.ToString(),
+                HostShutdownToken = exitToken
+            };
+        }
+
+        private void SetupNginx(string redirectUri, Uri originalUri)
+        {
+            // copy nginx.conf template and replace pertinent information
+            var configTemplateContent = File.ReadAllText(Path.Combine(DeploymentParameters.ApplicationPath, "nginx.conf"));
```

Comment:
For IIS Express we have different configs for different tests, specified via DeploymentParameters.ApplicationHostConfigTemplateContent/ApplicationHostConfigLocation. Rename that to something more generic (ServerConfigTemplate/ServerConfigLocation?) and re-use it here.